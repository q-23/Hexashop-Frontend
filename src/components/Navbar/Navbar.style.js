import Component from "components/_Shared/Component";

const NavbarStyle = Component('nav')`
	border-bottom: .1px solid #FFFFFF;
	background-color: #3750CF;
	overflow: hidden;
	position: sticky;
	z-index: 100;
	height: 3em;
	width: 100%;
	top: 0;
`;

export { NavbarStyle };