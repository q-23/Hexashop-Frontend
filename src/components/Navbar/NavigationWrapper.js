import Component from "../_Shared/Component";

const NavigationWrapper = Component('nav')`
	height: 3em;
	width: 100%;
	background-color: #3750CF;
	border-bottom: .1px solid #FFFFFF;
	overflow: hidden;
	position: sticky;
	top: 0;
	z-index: 100;
`;

export default NavigationWrapper;