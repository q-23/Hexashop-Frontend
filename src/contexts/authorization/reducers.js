const reducers = localStorageName => (state, action) => {
	switch (action.type) {
		case 'init': {
			return state;
		}
		case 'login': {
			const token = action.payload;
			localStorage.setItem(localStorageName, JSON.stringify(token));

			return token;
		}
		case 'logout': {
			localStorage.removeItem(localStorageName);
			return '';
		}
		default:
			return state;
	}
};

export default reducers;
