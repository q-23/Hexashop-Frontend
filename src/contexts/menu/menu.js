import React, {createContext, useEffect, useState} from 'react';
import { withRouter } from 'react-router-dom';
export const MenuContext = createContext();

export const StateProviderMenu = ({ children, location }) => {
	const [menuCategories, setMenuCategories] = useState([{ category_path: '/all_products', category_name: 'All products' }]);
	const [activeCategory, setActiveCategory] = useState(undefined);
	const [menuOpen, setMenuOpen] = useState(false)

	const fetchCategories = async () => {
		const categoriesBuffer = await fetch(`${process.env.REACT_APP_API_URL}/category`, {
			headers : {
				'Content-Type': 'application/json',
				'Accept': 'application/json'
			}
		});
		const categoriesResolved = await categoriesBuffer.json();
		const categoriesWithMappedPaths = categoriesResolved.map(e => ({ ...e, category_path: '/category' + e.category_path}));
		setMenuCategories([...menuCategories, ...categoriesWithMappedPaths]);
	}

	useEffect(() => {
		fetchCategories();
	//	eslint-disable-next-line
	}, []);

	useEffect(() => {
		const foundCategory = menuCategories.find(category => category.category_path.split('/')[2] === location.pathname.split('/')[2]);
		if (foundCategory) {
			setActiveCategory(foundCategory)
		}
	} ,[location.pathname, menuCategories])

	return (
		<MenuContext.Provider
			value={{
				menuOpen,
				setMenuOpen,
				menuCategories,
				setMenuCategories,
				activeCategory,
				setActiveCategory
			}}
		>
			{children}
		</MenuContext.Provider>
	);
};

export default withRouter(StateProviderMenu);