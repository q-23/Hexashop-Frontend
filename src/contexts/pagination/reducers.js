import paginationActions from './actions';

const paginationReducers = (state, action) => {
	const { type, payload } = action;
	switch (type) {
		case paginationActions.INIT: {
			return {
				skip: 0,
				limit: 10,
				numberOfPages: 1,
				numberOfItems: 0,
				currentPage: 1
			}
		}
		case paginationActions.SET_ITEMS_COUNT: {
			const itemsCount = payload;
			const { limit } = state;
			return { ...state, numberOfPages: Math.ceil(itemsCount/limit) };
		}
		case paginationActions.NEXT_PAGE: {
			if(state.currentPage < state.numberOfPages) {
				return { ...state, currentPage: state.currentPage + 1 };
			}
			return state;
		}
		case paginationActions.PREVIOUS_PAGE: {
			if(state.currentPage > 1) {
				return { ...state, currentPage: state.currentPage - 1 };
			}
			return state;
		}
		case paginationActions.SELECT_PAGE: {
			return { ...state, currentPage: payload };
		}
		case paginationActions.SET_SKIP_PAGES: {
			return { ...state, skip: state.currentPage * state.limit - state.limit };
		}
		default:
			return state;
	}
};

export default paginationReducers;
