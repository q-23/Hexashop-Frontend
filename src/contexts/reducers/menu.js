import React, {createContext, useEffect, useState} from 'react';
import { withRouter } from 'react-router-dom';
export const MenuContext = createContext();

export const StateProviderMenu = ({ children, location }) => {
	const [menuCategories, setMenuCategories] = useState([{ category_path: '/all_products', category_name: 'All products', showInMenu: true }]);
	const [activeCategory, setActiveCategory] = useState(undefined);
	const [menuOpen, setMenuOpen] = useState(false)

	useEffect(() => {
		const fetchCategories = async () => {
			const categoriesBuffer = await fetch(`${process.env.REACT_APP_API_URL}/category`, {
				headers : {
					'Content-Type': 'application/json',
					'Accept': 'application/json'
				}
			});
			const categoriesResolved = await categoriesBuffer.json();
			const categoriesWithShowInMenu = categoriesResolved.map(e => ({ ...e, showInMenu: true, category_path: '/category' + e.category_path}));
			setMenuCategories([...menuCategories, ...categoriesWithShowInMenu]);
		}
		fetchCategories();
	}, []);

	useEffect(() => {
		const foundCategory = menuCategories.find(category => category.category_path === location.pathname);
		if (foundCategory) {
			setActiveCategory(foundCategory)
		}
	} ,[location.pathname, menuCategories])

	return (
		<MenuContext.Provider
			value={{
				menuOpen,
				setMenuOpen,
				menuCategories,
				setMenuCategories,
				activeCategory,
				setActiveCategory
			}}
		>
			{children}
		</MenuContext.Provider>
	);
};

export default withRouter(StateProviderMenu);