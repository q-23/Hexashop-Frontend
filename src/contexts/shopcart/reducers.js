import shopcartActions from './actions';

const shopcartReducers = (state, action) => {
	const { type, payload } = action;
	switch (type) {
		case shopcartActions.ADD_ITEM_TO_CART: {
			const { product_id, quantity } = payload;
			if(!state.id) {
				return { ...state, products: { ...state.products, [product_id]: quantity } }
			}
			return { ...state, products: { ...state.products, [product_id]: state[product_id] + quantity } }
		}
		case shopcartActions.CALCULATE_TOTAL_QUANTITY: {
			const { products } = state;
			const productsCount = Object.values(products).reduce((acc, val) => {
				return acc + val
			}, 0)
			return { ...state, productsCount }
		}
		default:
			return state;
	}
};

export default shopcartReducers;
