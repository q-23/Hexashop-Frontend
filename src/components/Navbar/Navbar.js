import React, { useContext } from "react";

import { withRouter } from 'react-router-dom';
import { Link } from "react-router-dom";

import Button from "components/_Shared/Button";
import Icon from "components/_Shared/Icon";
import CartIcon from "components/CartIcon";

import { NavbarStyle } from "./Navbar.style";

import { MenuContext } from 'contexts/menu/menu';
import {useStateValueAuthorization} from "contexts/authorization/authorization";

const Navbar = () => {
	const { menuOpen, setMenuOpen } = useContext(MenuContext);
	const [auth] = useStateValueAuthorization();
	console.log(auth)
	return(
		<NavbarStyle>
			<Button navbar_button nav_border_right onClick={() => setMenuOpen(!menuOpen)}>
				<Icon className={`fa fa-${menuOpen ? 'times' : 'bars'}`} size={'1.7em'} color={'#FFF'}/>
			</Button>
			<Button navbar_button  nav_border_right>
				<Icon className="fa fa-github" size={'1.7em'} color={'#FFF'}/>
			</Button>
			<Link to={!!auth && auth.length ? '/account' : '/login'}>
				<Button navbar_button float={'right'} nav_border_right>
					<Icon className="fa fa-user" size={'1.7em'} color={'#FFF'}/>
				</Button>
			</Link>
			<Button navbar_button float={'right'}  nav_border_left>
				<CartIcon />
			</Button>
		</NavbarStyle>
	)
};

export default withRouter(Navbar);