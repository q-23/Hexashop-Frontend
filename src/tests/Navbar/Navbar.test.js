import React, { useContext } from 'react';

import { render, fireEvent } from '@testing-library/react'
import {MenuContext} from "../../contexts/reducers/menu";
import '@testing-library/jest-dom/extend-expect'
import { renderWithRouter } from "../utils";

import Navbar from "../../components/Navbar/Navbar";

const TestComponent = () => {
	const { menuOpen } = useContext(MenuContext);

	return (
		<div>
			<Navbar/>
			TEKST
			{ menuOpen.toString() }
		</div>
	)
}

describe('[NAVBAR]', () => {
	test('Should render component', () => {
		const wrapper = render(
			renderWithRouter(<Navbar/>)
		);
		expect(wrapper).toMatchSnapshot();
	});

	test('Should open the menu after clicking proper button', () => {
		const wrapper = render(
			renderWithRouter(<TestComponent/>)
		);
		fireEvent.click(wrapper.container.querySelector('.fa-bars'))
		expect(wrapper.getByText(/true/)).toBeInTheDocument();
	});

	test('Should close the menu after clicking button two times', () => {
		const wrapper = render(
			renderWithRouter(<TestComponent/>)
		);
		fireEvent.click(wrapper.container.querySelector('.fa-bars'))
		expect(wrapper.getByText(/true/)).toBeInTheDocument();
		fireEvent.click(wrapper.container.querySelector('.fa-times'))
		expect(wrapper.getByText(/false/)).toBeInTheDocument();
	});
});

