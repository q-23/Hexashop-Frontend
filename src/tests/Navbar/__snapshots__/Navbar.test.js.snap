// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[NAVBAR] Should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="sc-AxgMl fSAtcU"
      >
        <button
          class="sc-AxirZ edtojf"
        >
          <i
            class="sc-AxiKw hemaxJ fa fa-bars"
            color="#FFF"
          />
        </button>
        <a
          href="https://github.com/q-23/Hexashop-Frontend"
          target="_blank"
        >
          <button
            class="sc-AxirZ edtojf"
          >
            <i
              class="sc-AxiKw hemaxJ fa fa-github"
              color="#FFF"
            />
          </button>
        </a>
        <a
          href="/login"
        >
          <button
            class="sc-AxirZ eqVsUa"
          >
            <i
              class="sc-AxiKw hemaxJ fa fa-user"
              color="#FFF"
            />
          </button>
        </a>
        <a
          href="/cart"
        >
          <button
            class="sc-AxirZ eqVsUa"
          >
            <div
              class="sc-AxhUy gjrxrs"
            >
              <i
                class="sc-AxiKw hemaxJ fa fa-shopping-basket"
                color="#FFF"
              />
              <span
                class="sc-AxhCb eGRRou"
              >
                0
              </span>
            </div>
          </button>
        </a>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="sc-AxgMl fSAtcU"
    >
      <button
        class="sc-AxirZ edtojf"
      >
        <i
          class="sc-AxiKw hemaxJ fa fa-bars"
          color="#FFF"
        />
      </button>
      <a
        href="https://github.com/q-23/Hexashop-Frontend"
        target="_blank"
      >
        <button
          class="sc-AxirZ edtojf"
        >
          <i
            class="sc-AxiKw hemaxJ fa fa-github"
            color="#FFF"
          />
        </button>
      </a>
      <a
        href="/login"
      >
        <button
          class="sc-AxirZ eqVsUa"
        >
          <i
            class="sc-AxiKw hemaxJ fa fa-user"
            color="#FFF"
          />
        </button>
      </a>
      <a
        href="/cart"
      >
        <button
          class="sc-AxirZ eqVsUa"
        >
          <div
            class="sc-AxhUy gjrxrs"
          >
            <i
              class="sc-AxiKw hemaxJ fa fa-shopping-basket"
              color="#FFF"
            />
            <span
              class="sc-AxhCb eGRRou"
            >
              0
            </span>
          </div>
        </button>
      </a>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
